src/relax/op/tensor/transform.cc

```diff
@@ -0,0 +1,441 @@
+  for (int i = 0; i < n_new_dim; ++i) {
+    output_shape[axes[i]] = IntImm(DataType::Int(64), 1);
+  }
+
+  int i_data_shape = 0;
+  for (int i = 0; i < output_ndim; ++i) {
+    if (output_shape[i].defined()) {
+      continue;
+    }
+    ICHECK_LT(i_data_shape, data_sinfo->ndim);
+    output_shape[i] = data_shape->values[i_data_shape];
+    ++i_data_shape;
+  }
+  ICHECK_EQ(i_data_shape, data_sinfo->ndim);
+  return TensorStructInfo(ShapeExpr(output_shape), data_sinfo->dtype);
+}
+
+TVM_REGISTER_OP("relax.expand_dims")
+    .set_num_inputs(1)
+    .set_attrs_type<ExpandDimsAttrs>()
+    .add_argument("data", "Tensor", "The input tensor.")
+    .set_attr<FInferStructInfo>("FInferStructInfo", InferStructInfoExpandDims);
+
+/* relax.squeeze */
+TVM_REGISTER_NODE_TYPE(SqueezeAttrs);
+
+Expr MakeSqueeze(Expr data, Optional<Array<Integer>> axis) {
+  ObjectPtr<SqueezeAttrs> attrs = make_object<SqueezeAttrs>();
+  attrs->axis = std::move(axis);
+
+  static const Op& op = Op::Get("relax.squeeze");
+  return Call(op, {std::move(data)}, Attrs{attrs}, {});
+}
+
+TVM_REGISTER_GLOBAL("relax.op.squeeze").set_body_typed(MakeSqueeze);
+
+StructInfo InferStructInfoSqueeze(const Call& call, const BlockBuilder& ctx) {
+  TensorStructInfo data_sinfo = GetUnaryInputTensorStructInfo(call, ctx);
+  const auto* attrs = call->attrs.as<SqueezeAttrs>();
+  if (attrs->axis.defined() && attrs->axis.value().empty()) {
+    return data_sinfo;
+  }
+
+  if (data_sinfo->IsUnknownNdim()) {
+    return TensorStructInfo(data_sinfo->dtype, kUnknownNDim);
+  }
+
+  Optional<Array<PrimExpr>> shape_value;
+  if (data_sinfo->shape.defined()) {
+    shape_value = Downcast<ShapeStructInfo>(data_sinfo->shape.value()->struct_info_)->values;
+  }
+
+  std::vector<bool> axis_removal_mask;
+  axis_removal_mask.resize(data_sinfo->ndim, /*value=*/false);
+
+  if (attrs->axis.defined()) {
+    std::vector<int> axes =
+        CheckAxesInRangeNonRepetitive(call, ctx, data_sinfo->ndim, attrs->axis.value());
+
+    if (!shape_value.defined()) {
+      return TensorStructInfo(data_sinfo->dtype, data_sinfo->ndim - axes.size());
+    }
+    for (int i = 0; i < static_cast<int>(axes.size()); ++i) {
+      // When `axis` is given, the dim length at the axes must be static constant integer 1.
+      const auto* int_len = shape_value.value()[axes[i]].as<IntImmNode>();
+      if (int_len == nullptr || int_len->value != 1) {
+        ctx->ReportFatal(Diagnostic::Error(call)
+                         << "Squeeze expects the input tensor shape values at the given axis "
+                            "positions to be all 1. However, the tensor shape at axis "
+                         << axes[i] << " is " << shape_value.value()[axes[i]]
+                         << " which is not 1. If it is symbolic, please use MatchCast to cast it "
+                            "to 1 before doing Squeeze.");
+      }
+      axis_removal_mask[axes[i]] = true;
+    }
+  } else {
+    if (!shape_value.defined()) {
```

Comment:
Add comment, squeeze all 1s

Reply:
Add a comment: this is a less well-defined path in Array API standard's squeeze, consider discourage usage later

Reply:
https://data-apis.org/array-api/latest/API_specification/generated/array_api.squeeze.html#array_api.squeeze